x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/rescorer/InferenceRescorer.java

```diff
@@ -0,0 +1,150 @@
+
+        // First take top slice of incoming docs, to be rescored:
+        TopDocs topNFirstPass = topN(topDocs, rescoreContext.getWindowSize());
+        // Save doc IDs for which rescoring was applied to be used in score explanation
+        Set<Integer> topNDocIDs = Arrays.stream(topNFirstPass.scoreDocs).map(scoreDoc -> scoreDoc.doc).collect(toUnmodifiableSet());
+        rescoreContext.setRescoredDocs(topNDocIDs);
+        ScoreDoc[] hits = topNFirstPass.scoreDocs.clone();
+        Arrays.sort(hits, Comparator.comparingInt(a -> a.doc));
+        SearchLookup sourceLookup = ltrRescoreContext.executionContext.lookup();
+        int hitUpto = 0;
+        int readerUpto = -1;
+        int endDoc = 0;
+        int docBase = 0;
+        List<LeafReaderContext> leaves = ltrRescoreContext.executionContext.searcher().getIndexReader().leaves();
+        LeafReaderContext currentSegment = null;
+        boolean changedSegment = true;
+        List<Map<String, Object>> docFeatures = new ArrayList<>(topNDocIDs.size());
+        int featureSize = ltrRescoreContext.valueFetcherList.size();
+        while (hitUpto < hits.length) {
+            final ScoreDoc hit = hits[hitUpto];
+            final int docID = hit.doc;
+            while (docID >= endDoc) {
+                readerUpto++;
+                currentSegment = leaves.get(readerUpto);
+                endDoc = currentSegment.docBase + currentSegment.reader().maxDoc();
+                changedSegment = true;
+            }
+            assert currentSegment != null : "Unexpected null segment";
+            if (changedSegment) {
+                // We advanced to another segment and update our document value fetchers
+                docBase = currentSegment.docBase;
+                for (InferenceRescorerContext.FieldValueFetcher vf : ltrRescoreContext.valueFetcherList) {
+                    vf.valueFetcher().setNextReader(currentSegment);
+                }
+                changedSegment = false;
+            }
+            int targetDoc = docID - docBase;
+            Map<String, Object> features = Maps.newMapWithExpectedSize(featureSize);
+            Source source = sourceLookup.getSource(currentSegment, targetDoc);
+            for (InferenceRescorerContext.FieldValueFetcher vf : ltrRescoreContext.valueFetcherList) {
+                features.put(vf.fieldName(), vf.valueFetcher().fetchValues(source, targetDoc, new ArrayList<>()).get(0));
+            }
+            docFeatures.add(features);
+            hitUpto++;
+        }
+        for (int i = 0; i < hits.length; i++) {
+            Map<String, Object> features = docFeatures.get(i);
+            try {
+                hits[i].score = ((Number) definition.infer(features, RegressionConfigUpdate.EMPTY_PARAMS).predictedValue()).floatValue();
+            } catch (Exception ex) {
+                logger.warn("Failure rescoring doc...", ex);
+            }
+        }
+        if (rescoreContext.getWindowSize() < hits.length) {
```

Comment:
will this ever be executed considering that we before limited hit.length to rescoreContext.getWindowSize()

Reply:
> will this ever be executed considering that we before limited hit.length to rescoreContext.getWindowSize()?

@mayya-sharipova do we really make that limitation? I thought you could set `size=20` and `window_size=10` in this instance we only rerank the top ten results from the 20 hits returned.

Reply:
I know we increase `size` on the shard if `window_size > size`